sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
You are right! 

But I think it is reasonable to copy the data twice because in step 1.,
`init = check_array(init, dtype=np.float64, copy=True)` is used to make sure `KMeans.init` is a copy of the array provided by users.

And in step 2., `centers = np.array(init, dtype=X.dtype)` 
is used to make sure the `centers` we compute won't alter its argument `init`.